zipper/helper/requests.go

```diff
@@ -5,17 +5,114 @@ import (
 	"context"
 	"io"
 	"io/ioutil"
+	"net"
 	"net/http"
 	"net/url"
 	"sync/atomic"
 
 	"github.com/ansel1/merry"
 	"github.com/go-graphite/carbonapi/limiter"
+	"github.com/go-graphite/carbonapi/pkg/parser"
 	util "github.com/go-graphite/carbonapi/util/ctx"
 	"github.com/go-graphite/carbonapi/zipper/types"
 	"go.uber.org/zap"
 )
 
+func HttpCode(err error) int {
+	if err == nil {
+		return http.StatusOK
+	}
+	if _, ok := err.(*net.OpError); ok {
+		return http.StatusServiceUnavailable
+	}
+	if urlErr, ok := err.(*url.Error); ok {
+		if _, ok = urlErr.Err.(*net.OpError); ok {
+			return http.StatusServiceUnavailable
+		}
+	}
+	return http.StatusInternalServerError
+}
+
+func MergeHttpErrors(errors []merry.Error) (int, []string) {
+	returnCode := http.StatusNotFound
+	errMsgs := make([]string, 0)
+	for _, err := range errors {
+		var code int
+		c := merry.RootCause(err)
+		if c == nil {
+			c = err
+		}
+		code = merry.HTTPCode(c)
+
+		if merry.HTTPCode(c) == 404 || merry.Is(c, parser.ErrSeriesDoesNotExist) {
+			continue
+		}
+		errMsgs = append(errMsgs, c.Error())
+		if code == 400 {
+			// The 400 is returned on wrong requests, e.g. non-existent functions
+			returnCode = code
+		} else if returnCode == 404 {
+			// First error
+			returnCode = code
+		} else if code == 503 && (returnCode == 504 || returnCode == 502) {
+			returnCode = 503
+		} else if (code < 502 || code > 504) && (returnCode >= 502 || returnCode <= 504) {
```

Comment:
I'm not really following the logic here and what makes 503 so special for example (it looks like this `else if` can be merged with previous one that was for 403 and as well with the one that's for 400).

I'm also not so sure if returnCode will ever be changed to something other than 400 or 404. It looks like that all changes that could've potentially set it to 504 or 503 are guarded with `(returnCode == 504 || returnCode == 502)` or with `(reutrnCode >= 502 || returnCode <= 504)`.

Reply:
Yes, logic of error handling is a main question.
May be better - priority of 400, 403, 500 and other non-retryble errors instead of communication errors (502, 503, 504).
Becouse in this case query might be wrong, access denied (possible limits in graphite-clickhouse). We don't need retry this queries.